### PR TITLE
Cap the fused cross entropy chunk target so chunking stays active on large GPUs

### PR DESCRIPTION
**Pull request by danielhanchen**
## Summary

The fused cross entropy auto chunk-size path targets half of the free VRAM per chunk. On large-memory GPUs (e.g. B200 class) half the free pool is big enough that the chunk count rounds down to one, which materializes the full logits plus their float32 upcast in a single transient. At long sequence lengths with large vocabs that transient dominates peak memory (2x 10-17 GB at 32k tokens for 150k-260k vocabs).

## What this does

Caps the auto target at 4 GB so chunking stays active regardless of how much memory is free. Small GPUs are unaffected since their half-free value stays below the cap.

## Testing

- Two unit tests: chunk count stays above one on a mocked 180 GiB pool, and the multiplier is identical for 120 GiB vs 320 GiB pools once the cap dominates.
- Measured on Qwen3-30B-A3B and gemma MoE 32k-token training runs: identical tok/s, peak memory down 6.4 GB and 21.4 GB respectively.

### CODE REVIEW

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
Set `UNSLOTH_ALLOW_CPU=1` at the top of the test so it imports on a CPU-only runner, where device detection would otherwise raise before the test can run.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/fused_losses/cross_entropy_loss.py [hunk 1]
```diff
@@ -144,7 +144,9 @@ def _get_chunk_multiplier(vocab_size, target_gb = None):
         free, total = torch.xpu.mem_get_info(0) if DEVICE_TYPE == "xpu" else torch.cuda.mem_get_info(0)
         free_gb = free / 1024 / 1024 / 1024
         free_gb = free_gb * 0.5
-        target_gb = free_gb
+        # Cap per-chunk target: on very large GPUs half the free pool rounds to a
+        # single chunk, materializing full float32 logits and dominating peak memory.
+        target_gb = min(free_gb, 4.0)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/fused_losses/cross_entropy_loss.py:149]: Good catch. get_chunk_size now rounds the chunk count up to the next multiple of four rather than to nearest, so the 4-8 GiB band splits into 4 chunks (2 GiB each) instead of collapsing to one, while configs whose logits already fit the target stay at a single chunk.

File: tests/test_cross_entropy_chunk_cap.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+os.environ.setdefault("UNSLOTH_ALLOW_CPU", "1")
+
+import importlib
+import types
+
+import pytest
+
+
+def _load_module(monkeypatch, free_bytes):
+    ce = importlib.import_module("unsloth_zoo.fused_losses.cross_entropy_loss")
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_cross_entropy_chunk_cap.py]: Fixed. The helper import is now wrapped so a zoo-only checkout skips the test instead of erroring during collection, and the full assertions still run wherever unsloth is installed.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
### Final results

Setup: fused cross entropy auto chunk-size target capped at 4 GB so chunking stays active on large-memory GPUs. Measured on 32k-token training runs; throughput unchanged.

| model | seq | tok/s | peak memory |
|---|---|---|---|
| Qwen3-30B-A3B | 32k | unchanged | -6.4 GB |
| gemma MoE | 32k | unchanged | -21.4 GB |

Takeaway: capping the target keeps the CE logits chunked on B200-class GPUs, cutting peak memory by up to 21.4 GB at 32k tokens with no throughput change. Small GPUs are unaffected.
